File: Leetcode/Burst Balloons/burst_balloons.py
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        # Append a balloon painted with `1` in the head, and a balloon painted with `1` in the tail:
        nums = [1] + nums + [1]
        # current balloons (with two `vitural` ballons)
        n = len(nums)
        # dp[i][j]: the maximum coins we can collect by bursting balloons from index `i` to index `j`
        dp = [[0 for _ in range(n)] for _ in range(n)]

        # compute dp[i][j]
        def cal(i, j):
            # memory
            if dp[i][j]:
                return dp[i][j]
            # Meaningless case. Just return 0. 
            elif i > j:
                return dp[i][j]
            else:
                coins_max = 0
                # loop through (j+1-i) possible positions for the last balloon
                for k in range(i, j + 1):
                    left = nums[i-1]
                    right = nums[j+1]
                    coins_k_max = cal(i, k - 1) + left * nums[k] * right + cal(k + 1, j)
                    coins_max = max(coins_max, coins_k_max)
                # update the memory
                dp[i][j] = coins_max
                return dp[i][j]
        # compute dp[1][n-2]
        return cal(1, n - 2)

File: Leetcode/Burst Balloons/burst balloons.cpp
int maxCoins(vector<int>& nums) {
    int N = nums.size();
    nums.insert(nums.begin(), 1);
    nums.insert(nums.end(), 1);

    // rangeValues[i][j] is the maximum # of coins that can be obtained
    // by popping balloons only in the range [i,j]
    vector<vector<int>> rangeValues(nums.size(), vector<int>(nums.size(), 0));
    
    // build up from shorter ranges to longer ranges
    for (int len = 1; len <= N; ++len) {
        for (int start = 1; start <= N - len + 1; ++start) {
            int end = start + len - 1;
            // calculate the max # of coins that can be obtained by
            // popping balloons only in the range [start,end].
            // consider all possible choices of final balloon to pop
            int bestCoins = 0;
            for (int final = start; final <= end; ++final) {
                int coins = rangeValues[start][final-1] + rangeValues[final+1][end]; // coins from popping subranges
                coins += nums[start-1] * nums[final] * nums[end+1]; // coins from final pop
                if (coins > bestCoins) bestCoins = coins;
            }
            rangeValues[start][end] = bestCoins;
        }
    }
    return rangeValues[1][N];
}
